### PR TITLE
Changes for August release of the C/C++ extensions

### PR DESCRIPTION
**Pull request by edumunoz**
@Andrew-MSFT, @pieandcakes, @jacdavis, @gregvanl

Please review.

### CODE REVIEW

File: docs/languages/cpp.md [hunk 1]
```diff
@@ -10,13 +10,13 @@ MetaDescription: Find out how to get the best out of Visual Studio Code and C++.
 
 # C/C++ for VS Code (Preview)
 
-C/C++ support for Visual Studio Code is provided today as a preview of our work to enable cross-platform C and C++ development using VS Code on Windows, Linux, and OS X. Our focus in this preview release is code editing and navigation support for C and C++ code everywhere that VS Code runs, as well as debugging on Linux, OS X, and Windows (GDB only with Cygin and MinGW).
+C/C++ support for Visual Studio Code is provided today as a preview of our work to enable cross-platform C and C++ development using VS Code on Windows, Linux, and OS X. Our focus in this preview release is code editing and navigation support for C and C++ code everywhere that VS Code runs, as well as debugging on Linux, OS X, and Windows.
```
Comment (Andrew-MSFT) [inline docs/languages/cpp.md]: Just came from a marketing meeting, they are changing the rules and we aren't supposed to use "VS Code" any longer.  Either "Visual Studio Code" or "Code".  Given that we have now added full support for Windows I would rephrase the last sentence to:  Our focus in this preview release is code editing, navigation, and debugging support for C and C++ code everywhere that VS Code runs

Comment (edumunoz) [inline docs/languages/cpp.md]: Sounds good to me. Are we also changing all the headings in this document? e.g.,

`# C/C++ for VS Code (Preview)` --> `# C/C++ for Visual Studio Code (Preview)`

File: docs/languages/cpp.md [hunk 2]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
+After you have set up the basics of your debugging environment as specified in [Getting Started](#getting_started), you can learn more details about debugging C/C++ here.
+
 VS Code supports the following debuggers for C/C++ depending on the operating system you are using:
 
-* **Linux**: supports debugging using GDB
-* **OS X**: supports using LLDB or GDB
-* **Windows**: currently supports GDB only (using Cygwin or MinGW)
+* **Linux**: using GDB
+* **OS X**: using LLDB or GDB
+* **Windows**: using the Visual Studio Windows Debugger or GDB (using Cygwin or MinGW)
```
Comment (Andrew-MSFT) [inline docs/languages/cpp.md]: I would drop the "using" and just say "GDB", "LLDB or GDB", "Visual Studio Windows Debugger or GDB (using Cygwin or MinGW)"

File: docs/languages/cpp.md [hunk 3]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
+The C/C++ extension for VS Code has the ability to debug multi-threaded programs. All threads and their call stacks appear in the __Call Stack__ section:
 
-### Core Dump debugging
+![Multi-threaded process](images/cpp/threads.png)
 
-The C/C++ extension for VS Code also has the ability to debug using a memory dump. To debug using a memory dump, open your launch.json file for editing and add the `coreDumpPath` property to the __C++ Launch__ configuration, setting its value to be a string containing the path to the core dump. This will even work for multi-threaded programs and x86 programs being debugged on an x64 machine.
+### Memory Dump Debugging
 
-### GDB and MI commands
+The C/C++ extension for VS Code also has the ability to debug using a memory dump. To debug using a memory dump, open your `launch.json` file for editing and add the `coreDumpPath` (for GDB or LLDB) or `dumpPath` (for the Visual Studio Windows Debugger) property to the __C++ Launch__ configuration, setting its value to be a string containing the path to the memory dump. This will even work for multi-threaded programs and x86 programs being debugged on an x64 machine.
```
Comment (Andrew-MSFT) [inline docs/languages/cpp.md]: -"also has the ability to debug a memory dump" (don't think we need the "using"
Is there value in calling out "even work for multi-threaded programs" here?

Comment (edumunoz) [inline docs/languages/cpp.md]: I agree on both. That's how it was originally written and I just changed some words to generalize it to Windows and Unix dumps.

File: docs/languages/cpp.md (hunk 1 above)
Comment (Andrew-MSFT) [inline docs/languages/cpp.md]: We should

Comment (gregvanl) [inline docs/languages/cpp.md]: This is still in discussion so please stick with the current convention: "Visual Studio Code" for the first use in a doc, then "VS Code" for later uses.  Please don't use "Code" to refer to the product.
If this really needs to change in the future, I'll take care of updating the docs. Thanks

File: docs/languages/cpp.md [hunk 4]
```diff
@@ -61,64 +61,13 @@ For more information on tasks, see [Integrate with External Tools via Tasks](/do
...
-            "windows": {
-                "MIMode": "gdb"
-            }
-        }
-    ]
-}
-```
+If you are debugging with GDB on Windows, see [Windows Debugging on Cygwin/MinGW](#debug_windows).
 
 ## Editing code
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: Capital C

File: docs/languages/cpp.md [hunk 5]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
 
 ## Debugging
 
+After you have set up the basics of your debugging environment as specified in [Getting Started](#getting_started), you can learn more details about debugging C/C++ here.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: Is "here" supposed to be a link?

File: docs/languages/cpp.md (hunk 2 above)
Comment (pieandcakes) [inline docs/languages/cpp.md]: +1

File: docs/languages/cpp.md (hunk 5 above)
Comment (edumunoz) [inline docs/languages/cpp.md]: No. Maybe change it to "this section"?

File: docs/languages/cpp.md [hunk 6]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
 
-* **Linux**: supports debugging using GDB
-* **OS X**: supports using LLDB or GDB
-* **Windows**: currently supports GDB only (using Cygwin or MinGW)
+* **Linux**: using GDB
+* **OS X**: using LLDB or GDB
+* **Windows**: using the Visual Studio Windows Debugger or GDB (using Cygwin or MinGW)
 
-### Windows debugging (Cygwin/MinGW) <a name="debug_windows"></a>
+### Windows Debugging on Cygwin/MinGW <a name="debug_windows"></a>
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: Windows Debugging with GDB on Cygwin/MinGW

Comment (pieandcakes) [inline docs/languages/cpp.md]: Is there a section for Visual Studio Windows Debugger also? Maybe rename the anchors too

File: docs/languages/cpp.md [hunk 7]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
-    "miDebuggerPath":"c:\\mingw\\bin\\gdb.exe"
+    "miDebuggerPath": "c:\\mingw\\bin\\gdb.exe"
 ```
 
 Cygwin/MinGW debugging on Windows supports both attach and launch debugging scenarios.
 
 ### Conditional Breakpoints
 
-Conditional breakpoints enable you to break execution on a particular line of code only when the value of the conditional is true. To set a conditional breakpoint, right-click on an existing breakpoint and select __Edit Breakpoint__, this opens a small peek window where you can enter the condition that must evaluate to true in order for the breakpoint to activate and break execution.
+Conditional breakpoints enable you to break execution on a particular line of code only when the value of the condition is true. To set a conditional breakpoint, right-click on an existing breakpoint and select __Edit Breakpoint__, this opens a small peek window where you can enter the condition that must evaluate to true in order for the breakpoint to activate and break execution.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: __Edit Breakpoint__. This opens a small peek....

File: docs/languages/cpp.md (hunk 6 above)
Comment (edumunoz) [inline docs/languages/cpp.md]: Renamed the anchors and title.

No section for VSWD because there isn't anything special to mention about it.

File: docs/languages/cpp.md (hunk 7 above)
Comment (pieandcakes) [inline docs/languages/cpp.md]: in order for the breakpoint to be hit during debugging.

File: docs/languages/cpp.md [hunk 8]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
 
 ![A conditional break](images/cpp/condbreak.png)
 
 In the editor, conditional breakpoints are indicated by a breakpoint symbol that has a black equals sigh inside of it. You can place the cursor over a conditional breakpoint to show its condition.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: sign

File: docs/languages/cpp.md [hunk 9]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
 
-### Function breakpoints
+### Function Breakpoints
 
-Function breakpoints enable you to break execution at the beginning of a function instead of on a particular line of code. To set a function breakpoint, on the __Debug Panel__, right click inside the __Breakpoints__ pane, then choose __Add Function Breakpoint__ and enter the name of the function on which you want to break execution.
+Function breakpoints enable you to break execution at the beginning of a function instead of on a particular line of code. To set a function breakpoint, on the __Debug__ pane, right click inside the __Breakpoints__ section, then choose __Add Function Breakpoint__ and enter the name of the function on which you want to break execution.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: no comma after "To set a function breakpoint "

Comment (edumunoz) [inline docs/languages/cpp.md]: why not

File: docs/languages/cpp.md [hunk 10]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
 
-### Expression evaluation
+### Expression Evaluation
 
 VS Code supports expression evaluation in several contexts:
 
-* You can type an expression into the __Watch__ pane and it will be evaluated each time a breakpoint is hit.
+* You can type an expression into the __Watch__ section and it will be evaluated each time a breakpoint is hit.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: __Watch__ section of the __Debug__ pane (?) might be clearer

File: docs/languages/cpp.md (hunk 3 above)
Comment (pieandcakes) [inline docs/languages/cpp.md]: please also remove "for editing"

File: docs/languages/cpp.md (hunk 5 above)
Comment (pieandcakes) [inline docs/languages/cpp.md]: more details about debugging C\C++ in this section?

File: docs/languages/cpp.md (hunk 9 above)
Comment (pieandcakes) [inline docs/languages/cpp.md]: how about the one after "pane"?

File: docs/languages/cpp.md [hunk 11]
```diff
@@ -194,57 +143,77 @@ To go to a symbol's definition, place your cursor on the symbol anywhere its use
...
+
+For example:
+
+```json
+    "additionalSOLibSearchPath": "/path/to/symbols:/another/path/to/symbols"
+```
+or
+
+```json
+    "symbolSearchPath": "C:\\path\\to\\symbols;C:\\another\\path\\to\\symbols"
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: can you add an example where the path has a space and you need to escape the quotes?

File: docs/languages/cpp.md [hunk 12]
```diff
@@ -253,19 +222,23 @@ You can execute GDB or MI commands directly through the debug console with the `
...
 
 All platforms:
 
-* Because the extension doesn't parse function bodies, Peek Definition and Go to Definition don't work for symbols defined inside the body of a function. 
+* Because the extension doesn't parse function bodies, Peek Definition and Go to Definition don't work for symbols defined inside the body of a function.
 
 ### Debugging
 
+All platforms:
+* There is no way to specify the source file mappings from the compiled locations to the current location. This prevents debugging an application in a different filesystem structure than where it was compiled.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: layout, not structure

File: docs/languages/cpp.md [hunk 13]
```diff
@@ -275,10 +248,13 @@ Linux:
 
 OS X:
 
-* Additional install steps need to be completed manually to use GDB on OS X. See _Manual Installation for the C++ Debugger extension_ in the [README](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools).
-* When attaching to a process with GDB, the application being debugged cannot be interrupted. GDB will only bind breakpoints set while the application is not running (either before attaching to the application, or while the application is in a stopped state). This is due to [a bug in GDB](https://sourceware.org/bugzilla/show_bug.cgi?id=20035).
-* Core dumps cannot be loaded when debugging with GDB because GDB [does not support the core dump format used in OS X](https://www.sourceware.org/ml/gdb/2014-01/msg00036.html).
-* When attached to a process with GDB, break-all will end the process.
+* LLDB:
+    * When debugging with LLDB, if the Terminal window is closed while in break mode, debugging does not stop. Debugging can be stopped by pressing the **Stop** button.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: add another one that when debugging is stopped, that the terminal window is also not closed.

File: docs/languages/cpp.md [hunk 14]
```diff
@@ -275,10 +248,13 @@ Linux:
...
-* When attaching to a process with GDB, the application being debugged cannot be interrupted. GDB will only bind breakpoints set while the application is not running (either before attaching to the application, or while the application is in a stopped state). This is due to [a bug in GDB](https://sourceware.org/bugzilla/show_bug.cgi?id=20035).
-* Core dumps cannot be loaded when debugging with GDB because GDB [does not support the core dump format used in OS X](https://www.sourceware.org/ml/gdb/2014-01/msg00036.html).
-* When attached to a process with GDB, break-all will end the process.
+* LLDB:
+    * When debugging with LLDB, if the Terminal window is closed while in break mode, debugging does not stop. Debugging can be stopped by pressing the **Stop** button.
+* GDB:
+    * Additional install steps need to be completed manually to use GDB on OS X. See _Manual Installation for the C++ Debugger extension_ in the [README](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools).
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: Additional manual install steps need to be completed to use GDB on OS X.

Comment (pieandcakes) [inline docs/languages/cpp.md]: please verify that the section name in the readme is still "Manual Installation..."

File: docs/languages/cpp.md [hunk 15]
```diff
@@ -290,10 +266,6 @@ Read on to find out about:
 
 ## Common Questions
 
-**Q: I can't debug multi-threaded code.**
-
-**A:** For multi-threaded debugging, you must use the version 1.1.1 of the C/C++ Extension for Visual Studio Code.
-
 **Q: My project won't load.**
 
 **A:** VS Code doesn't currently support C++ project files, instead it considers a directory of your choosing to be the workspace of your project. Source code files inside that directory and its sub-directories are part of the workspace.
```
Comment (pieandcakes) [inline docs/languages/cpp.md]: Please add a section like:
Q: Where do I report issues or get help?
A: Please file an issue one [here] the GitHub link

File: docs/languages/cpp.md [hunk 16]
```diff
@@ -253,19 +222,23 @@ You can execute GDB or MI commands directly through the debug console with the `
...
 
 All platforms:
 
-* Because the extension doesn't parse function bodies, Peek Definition and Go to Definition don't work for symbols defined inside the body of a function. 
+* Because the extension doesn't parse function bodies, Peek Definition and Go to Definition don't work for symbols defined inside the body of a function.
 
 ### Debugging
 
+All platforms:
+* There is no way to specify the source file mappings from the compiled locations to the current location. This prevents debugging an application in a different filesystem layout than where it was compiled.
```
Comment (gregvanl) [inline docs/languages/cpp.md]: I think "file system" is more commonly used rather than "filesystem".
